lib/diagnostics_channel.js

```diff
@@ -138,10 +226,170 @@ function hasSubscribers(name) {
+      assertChannel(end, 'nameOrChannels.end');
+      assertChannel(asyncStart, 'nameOrChannels.asyncStart');
+      assertChannel(asyncEnd, 'nameOrChannels.asyncEnd');
+      assertChannel(error, 'nameOrChannels.error');
+
+      this.start = start;
+      this.end = end;
+      this.asyncStart = asyncStart;
+      this.asyncEnd = asyncEnd;
+      this.error = error;
+    } else {
+      throw new ERR_INVALID_ARG_TYPE('nameOrChannels',
+                                     ['string', 'object', 'Channel'],
+                                     nameOrChannels);
+    }
+  }
+
+  subscribe(handlers) {
+    for (const name of traceEvents) {
+      if (!handlers[name]) continue;
+
+      this[name]?.subscribe(handlers[name]);
+    }
+  }
+
+  unsubscribe(handlers) {
+    let done = true;
+
+    for (const name of traceEvents) {
+      if (!handlers[name]) continue;
+
+      if (!this[name]?.unsubscribe(handlers[name])) {
+        done = false;
+      }
+    }
+
+    return done;
+  }
+
+  traceSync(fn, ctx = {}, thisArg, ...args) {
+    const { start, end, error } = this;
+
+    try {
+      const result = start.runStores(ctx, fn, thisArg, ...args);
+      ctx.result = result;
+      return result;
+    } catch (err) {
+      ctx.error = err;
+      error.publish(ctx);
+      throw err;
+    } finally {
+      end.publish(ctx);
+    }
+  }
+
+  tracePromise(fn, ctx = {}, thisArg, ...args) {
+    const { start, end, asyncStart, asyncEnd, error } = this;
+
+    function reject(err) {
+      ctx.error = err;
+      error.publish(ctx);
+      asyncStart.publish(ctx);
+      // TODO: Is there a way to have asyncEnd _after_ the continuation?
+      asyncEnd.publish(ctx);
+      return PromiseReject(err);
+    }
+
+    function resolve(result) {
+      ctx.result = result;
+      asyncStart.publish(ctx);
+      // TODO: Is there a way to have asyncEnd _after_ the continuation?
+      asyncEnd.publish(ctx);
+      return result;
+    }
+
+    try {
+      let promise = start.runStores(ctx, fn, thisArg, ...args);
+      // Convert thenables to native promises
+      if (!(promise instanceof Promise)) {
+        promise = PromiseResolve(promise);
+      }
+      return PromisePrototypeThen(promise, resolve, reject);
+    } catch (err) {
+      ctx.error = err;
+      error.publish(ctx);
+      throw err;
+    } finally {
+      end.publish(ctx);
+    }
+  }
+
+  traceCallback(fn, position = 0, ctx = {}, thisArg, ...args) {
+    const { start, end, asyncStart, asyncEnd, error } = this;
+
+    function wrappedCallback(err, res) {
+      if (err) {
+        ctx.error = err;
+        error.publish(ctx);
+      } else {
+        ctx.result = res;
+      }
+
+      asyncStart.publish(ctx);
+      try {
+        if (callback) {
+          return ReflectApply(callback, this, arguments);
+        }
+      } finally {
+        asyncEnd.publish(ctx);
+      }
+    }
+
+    const callback = ArrayPrototypeAt(args, position);
+    if (typeof callback === 'function') {
```

Comment:
if `callback` is not a function a start event will be emitted but never an end. This likely results in problems for consumers.

I think we should either skip the start and swallow this wrong usage or throw an ` ERR_INVALID_ARG_TYPE`.

Reply:
The `end` event _will_ run, it's `asyncStart` and `asyncEnd` that will not. It will just look like a sync function if it can't find a callback.

Reply:
Ah right. But wouldn't a consumer "wait" on asyncEnd in this case as start came on a channel intended for async?

Reply:
Well, it depends on the use, but the _intent_ is for spans to get stored in AsyncLocalStorage, in which case they'll just go out of scope when the descending async does, so not really any "cleanup" to worry about. I _do_ see the value in handling improper use better though. 🤔

Reply:
The cleanup is more on consumer side. Not the Async Local part.
I think more about an API like `function fooDbExecute(query, cb)` which usually emits `fooDb.execute.start`, `fooDb.execute.end` and `fooDb.execute.asyncEnd`. A diag channel consomer would start a span in start event and end it ein asyncEnd event as this represents the duration. such a span is never ended if `cb` is not given as someone does a fire and forget request.
Db driver publishing to diag channel should use a dummy callback in this case to fix this.

Reply:
Yep, that's one of those "depends on the use" situations, but I would for sure like to avoid enabling footguns. Do you think just throwing an error when a callback was not found at the location implied by the caller is reasonable? It's definite user-error when telling it the callback position is N and it's not there. It _could_ leak out of the function itself though to the caller of whatever is calling the TracingChannel interface, if they don't validate their inputs sufficiently, but maybe that's fine?

Reply:
I think throwing is fine in this case as it is a clear wrong use. User should either switch to traceSync or supply a callback. Hiding this wrong use will likely not help anyone.
If user complain we can change it anyway later.

Reply:
I changed it to throw now. 👍🏻